core/vm/contracts.go

```diff
@@ -533,3 +544,59 @@ func (c *vrf) Run(input []byte) ([]byte, error) {
 	// So here we simply return it
 	return append([]byte{}, input...), nil
 }
+
+// SHA3-256 FIPS 202 standard implemented as a native contract.
+
+type sha3fip struct{}
+
+// TODO Check if the gas price calculation needs modification
+// RequiredGas returns the gas required to execute the pre-compiled contract.
+//
+// This method does not require any overflow checking as the input size gas costs
+// required for anything significant is so high it's impossible to pay for.
+func (c *sha3fip) RequiredGas(input []byte) uint64 {
+	return uint64(len(input)+31)/32*params.Sha256PerWordGas + params.Sha256BaseGas
+}
+func (c *sha3fip) Run(input []byte) ([]byte, error) {
+	hexStr := common.Bytes2Hex(input)
+	pub, _ := hex.DecodeString(hexStr)
+	h := sha3.Sum256(pub[:])
+	return h[:], nil
+}
+
+// ECRECOVER implemented as a native contract.
+type ecrecoverPublicKey struct{}
+
+func (c *ecrecoverPublicKey) RequiredGas(input []byte) uint64 {
+	return params.EcrecoverGas
+}
+
+func (c *ecrecoverPublicKey) Run(input []byte) ([]byte, error) {
+	const ecrecoverPublicKeyInputLength = 128
```

Comment:
Are these code from a existing implementation or it's a new implementation by yourself? If it's existing one, can you add the link to the original code?

Reply:
ecrecover public key function is similar to the "ecrecover" which retrieves the address while the former has one step lesser, converting the public key into an address.

but the sha3fips code implementation is our own using the "golang.org/x/crypto/sha3" crypto library.

Reply:
ok, looks good.